File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        open_max = 0
        open_min = 0

        for ch in s:
            if ch == '(':
                open_max += 1
                open_min += 1
            elif ch == ')':
                open_max -= 1
                open_min -= 1
            else:
                open_max += 1    # if * becomes '('
                open_min -= 1    # if * becomes ')'

            open_min = max(open_min, 0)   # open_min cannot be negative
            if open_max < 0:             # // not enough '(' to close parenthesis
                return False

        return open_min == 0

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.java
public class Solution {
    public boolean checkValidString(String s) {
        int openMax = 0, openMin = 0;
        for(var ch : s.toCharArray()){
            if(ch == '('){
                openMax++;
                openMin++;
            }
            else if(ch == ')'){
                openMax--;
                openMin--;
            }
            else if(ch == '*'){
                openMax++;  // if * becomes '('
                openMin--;  // if * becomes ')'
            }

            openMin = Math.max(openMin, 0); // openMin cannot be negative
            if(openMax < 0) return false;   // not enough '(' to close parenthesis
            
        }

        return openMin == 0;
    }
}
